### PR TITLE
Add zero count outcomes

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
The ability to compute the counts of outcomes from raw samples has recently been added to PennyLane (https://github.com/PennyLaneAI/pennylane/pull/2876 https://github.com/PennyLaneAI/pennylane/pull/2686 https://github.com/PennyLaneAI/pennylane/pull/2839).

The output of the new measurement is a dictionary where keys are the outcomes and the values are counts. This PR adds the option to include possible outcomes that were not observed as entries in the dictionary, with value 0.

**Description of the Change:**
Adds the kwarg `all_outcomes` to the `counts` function, with a default value of False. Calling `counts(all_outcomes=True)` can be used to return all possible outcomes, including unobserved outcomes. 

Uses the number of wires measured (if counting outcomes of measured computational basis states) or eigenvalues of observable to determine all possible outcomes, and populates a dictionary with all values set to 0, then updates with counts where relevant.

The variable `all_outcomes` is saved on the `MeasurementProcess` object that is returned by the `counts` function, as `MeasurementProcess.return_type.all_outcomes`. This variable can then be accessed by `_samples_to_counts` as `obs.return_type.all_outcomes` when creating the counts dictionary.

**Benefits:**
If `all_outcomes=True`, querying dictionary about a possible outcome that was not observed does not throw a KeyError, and printing the dictionary shows all possible outcomes. A KeyError is raised only if the user tries to access results for an outcome that does not match the possibilities for the measured system.

**Possible Drawbacks:**
None I can think of.

**Related GitHub Issues:**
(https://github.com/PennyLaneAI/pennylane/issues/2864)

### CODE REVIEW

**Comment by lillian542** [comment]
When I run 'make docs', I get the error xcrun: error: invalid active developer path. Suggested solutions for this online seem to include downloading a 15 GB program, and I only have 5GB of space left on my personal computer. Would it be possible for someone else to run the 'make docs' check?

**Comment by lillian542** [comment]
Most of the test pass, but there is an error when running 

`res = qml.execute_new(tapes=[qnode.tape], device=dev, gradient_fn=None)`

in line 904 of *test_new_return_types.py* that I'm not sure how to handle. 

I've gotten myself down a bit of a rabbit-hole trying to sort out what qml.execute is doing, how qnode.tape work, and what is being passed to the function as meas1 and meas2 when the test are run. Another set of eyes and/or a rundown of what this function is doing would be a big help.

**Comment by antalszava** [comment]
Hi @lillian542, thanks again!

For building the docs: we have a continuous integration check for building the docs, it's at the bottom of this PR. See a mention of it in the [CI checks section of our docs](https://pennylane.readthedocs.io/en/stable/development/guide/pullrequests.html#continuous-integration-checks).

**Comment by antalszava** [comment]
> Most of the test pass, but there is an error when running res = qml.execute_new(tapes=[qnode.tape], device=dev, gradient_fn=None) in line 904 of test_new_return_types.py that I'm not sure how to handle.
I've gotten myself down a bit of a rabbit-hole trying to sort out what qml.execute is doing, how qnode.tape work, and what is being passed to the function as meas1 and meas2 when the test are run. Another set of eyes and/or a rundown of what this function is doing would be a big help.

This test file is related to a major change that is experimental for now. If working it out doesn't come along, the marker `pytest.mark.xfail` can be put on the failing test cases. We're actively working on this feature and can resolve the issues if this is the only blocker. :+1: 

I'll leave a more detailed review on the PR today, thank you! :slightly_smiling_face:

**Comment by lillian542** [comment]
> If working it out doesn't come along, the marker `pytest.mark.xfail` can be put on the failing test cases. 

Alright, that's what I've done for now; I'm not sure quite what it going on, it seems to be passed an operator where `observable.name` is `['PauliZ', 'PauliZ']`, and calling `observable.compute_eigvals()` raises an `EigvalsUndefinedError`. That's as far as I've gotten. 

> I'll leave a more detailed review on the PR today, thank you! 🙂

Great, looking forward to hearing your feedback!

**Review by antalszava — COMMENTED** [review]
Hi @lillian542, thank you! Overall looking great, I had some questions.

While reading the PR, a major thing came to mind that was not mentioned in the issue. Looking at the concrete implementation and considering the cost of obtaining eigenvalues, it might be best to preserve the previous behaviour and add zero count outcomes as an optional feature. A keyword argument can be added to `qml.counts`, e.g., `all_outcomes` and making `all_outcomes=False` the default. Since this was not originally mentioned in the issue, we can go forward with this PR as is and the team can implement this kwarg logic (let me know).

Reasons for having a sparse dict by default (`all_outcomes=False`):
* `qml.eigvals` can be costly to use for sizable Hamiltonians;
* For many qubits, the length of the output for `qml.counts` grows significantly.

In return, having zero count outcomes (`all_outcomes=True`) is still great to have when users would like to see all outcomes in the dictionary (e.g., users who are learning quantum computing).

**Comment by lillian542** [comment]
Thanks @antalszava! That makes sense - I’m happy to implement that tomorrow afternoon. Alternatively, I could change the implementation to use `defaultdict`, as suggested in the issue. 

My reasoning for doing it this way was that it didn’t seem ideal that it would return the same result if you queried the dictionary about a state the system *couldn’t be in*, as it would for a result that would be *possible* but didn’t occur in the given measurement. 

It makes most sense to me that the former would return an error, and the latter a count of 0. But I can see that this implementation isn’t practical for larger systems, so if its a feature that would be generally appreciated, maybe the `defaultdict` solution is better than an optional, computation-heavy feature?

File: pennylane/_qubit_device.py
Comment (lillian542) [reply pennylane/_qubit_device.py:1466]: No, as far as I know it should be fine. I just added this at one point because I'm not very familiar with JAX and wasn't sure what the comments in the code about JAX compatibility were referring to, so I wanted to remember to check - and then forgot to remove it.

**Comment by lillian542** [comment]
It’s set up with an `all_outcomes` kwarg for the `counts` function now. 

I considered implementing it the same way `Counts` was implemented in PR [2686](https://github.com/PennyLaneAI/pennylane/pull/2686), as a new `ObservableReturnType` called `AllCounts`, but the two options don’t seem as meaningfully different as Counts vs Samples, and it seems excessive if a new `ObservabeReturnType` is needed every time optional features are added to measurements. Instead I opted to save the information as `MeasurementProcess.return_type.all_outcomes`, so that it can be accessed from the observable passed to `device._samples_to_counts`.

If there’s something I’ve missed that makes that a poor solution, I can switch the implementation to add an `ObservableReturnType` as was done for creating the counts option (or else I’m open to other suggestions). Let me know what you think!

File: tests/test_new_return_types.py
Comment (lillian542) [reply tests/test_new_return_types.py]: This test checks the default behaviour (`all_outcomes=False`), so it now passes. I've removed the `@pytest.mark.xfail` and the ToDo. Should a test with `counts(all_outcomes=True)` be added?

File: pennylane/_qubit_device.py
Comment (lillian542) [reply pennylane/_qubit_device.py]: I've tried testing this, but I'm not sure I tested the right thing. 

I've passed it the PauliZ operator 'manually' as `qml.Hermitian`, and that works (see `test_all_outcomes_hermitian` in _tests/test_measurements.py_). 

However, it returns the eigenvalues as floats, where the keys for the eigenvalues for the other operators are integers. But when returning basis states, the keys are strings, so its already not a consistent data type across the dictionaries, and it still functions the same, so maybe this isn't really an issue?

Is this a good enough test for `qml.Hermitian`?

File: tests/test_measurements.py
Comment (lillian542) [reply tests/test_measurements.py]: I've reverted these back to how they were previously, so that they just test the default implementation of `counts`. I've added new functions to test `counts(all_outcomes=True)`.

File: doc/introduction/measurements.rst
Comment (antalszava) [inline doc/introduction/measurements.rst:172]:
```suggestion
Per default, only observed outcomes are included in the dictionary. The kwarg ``all_outcomes=True`` can 
```

File: doc/introduction/measurements.rst
Comment (antalszava) [inline doc/introduction/measurements.rst:173]:
```suggestion
be used to display all possible outcomes, including those that were observed ``0`` times in sampling.
```

File: doc/introduction/measurements.rst
Comment (antalszava) [inline doc/introduction/measurements.rst:175]:
```suggestion
For example, we could run the previous circuit with ``all_outcomes=True``:
```

File: doc/introduction/measurements.rst
Comment (antalszava) [inline doc/introduction/measurements.rst:191]:
```suggestion
Note: For complicated Hamiltonians, this can add considerable overhead time (due to the cost of calculating 
```

File: doc/introduction/measurements.rst
Comment (antalszava) [inline doc/introduction/measurements.rst:192]:
```suggestion
eigenvalues to determine possible outcomes), and as the number of qubits increases, the length of the output 
```

File: doc/introduction/measurements.rst
Comment (antalszava) [inline doc/introduction/measurements.rst:195]:
```suggestion
```

File: pennylane/measurements.py
Comment (antalszava) [inline pennylane/measurements.py:757]:
```suggestion
    Executing this QNode shows only the observed outcomes:
```

File: pennylane/measurements.py
Comment (antalszava) [inline pennylane/measurements.py:695]: Note: when an observable is specified we have eigenvalues as keys.
```suggestion
            outcomes (default), or whether it will display all possible outcomes for the system
```

File: pennylane/measurements.py
Comment (antalszava) [inline pennylane/measurements.py:761]: Might be good to move the note
>     Passing all_outcomes=True will create a dictionary that displays all possible outcomes.

Just above the other example.
```suggestion

    .. code-block:: python3

    dev = qml.device("default.qubit", wires=2, shots=4)

    @qml.qnode(dev)
    def circuit():
        qml.PauliX(wires=0)
        return qml.counts()

    Executing this QNode shows only the observed state:

    >>> circuit()
    {'01': 4}

    Passing all_outcomes=True will create a dictionary that displays all possible outcomes.

```

File: tests/test_measurements.py
Comment (antalszava) [inline tests/test_measurements.py:870]: Let's also have a test case where we have multiple measurements including `qml.counts(..., all_outcomes=True)`.

**Review by antalszava — COMMENTED** [review]
Hi @lillian542, the latest changes are looking good! :+1: :tada: I think the current approach works well.

Once the comments are addressed, this PR will be close to merge-ready from my side. I'll ask for a second reviewer on the PR.

File: doc/introduction/measurements.rst
Comment (AlbertMitjans) [inline doc/introduction/measurements.rst:204]: Remove third qubit of the output of `print(result)`:

```pycon
[tensor(0.026, requires_grad=True) tensor(0.026, requires_grad=True)
 tensor({'01': 513, '11': 487}, dtype=object, requires_grad=True)]
```

Moreover, I get a tuple instead of a tensor as output. @antalszava was this changed in the return types PR?

File: pennylane/measurements.py
Comment (AlbertMitjans) [inline pennylane/measurements.py:801]: These setters are not very pythonic... we should change this if we can.

I can't think of a different implementation besides adding a new `ObservableReturnType`. @antalszava do you think it would be excessive to add a new enum type? If not we can keep this as it is.

**Review by AlbertMitjans — COMMENTED** [review]
Nice work! Added some comments

File: doc/introduction/measurements.rst
Comment (antalszava) [reply doc/introduction/measurements.rst:204]: No, it was reworked in https://github.com/PennyLaneAI/pennylane/pull/2839, but this output was missed there. Thanks for catching it!

File: pennylane/measurements.py
Comment (antalszava) [reply pennylane/measurements.py:801]: I think so, but we can definitely do that if we think it'll improve code quality. No issues with having e.g., `Counts` and `AllCounts`.

File: doc/introduction/measurements.rst
Comment (AlbertMitjans) [reply doc/introduction/measurements.rst:204]: @lillian542 could you change the output to a tuple too?

```pycon
(0.026, 0.026, {'01': 513, '11': 487})
```

File: pennylane/measurements.py
Comment (lillian542) [reply pennylane/measurements.py:801]: @AlbertMitjans, sure, I'll switch it to AllCounts and go through the other comments tomorrow.

File: pennylane/_qubit_device.py
Comment (lillian542) [reply pennylane/_qubit_device.py:1519]: @AlbertMitjans - I don't know if we do. The values in the standard Counts dictionary all have the type `np.int64`, so when initialising the dictionary as `outcome_dict = {k: 0 for k in outcomes}` you end up with mixed data types for the values. 

This caused an error at some point with a `dtype` check, but I don't remember in what context and I can't get it to reproduce the error now. Should I just remove it?

**Comment by lillian542** [comment]
Hi guys! It's implemented using `AllCounts` now, and I've gone through the comments. Could you give it another look over when you have time?

It would be great if you could quickly take an extra look at the tests, its not something I’ve done before so I’m not confident I’ve made good choices or covered all the bases that should be covered.

Also, to note, right now the keys for the dictionary returned by `counts` can be strings (for basis state measurements), integers (for operators like `PauliZ`), or floats (for `qml.Hermitian`). Is the lack of consistency a problem?

@antalszava @AlbertMitjans

**Review by AlbertMitjans — COMMENTED** [review]
Nice work!! 💯 

In the last commits to `master` there were some additions that require changes in this PR:

- Add `qml.measurements.AllCounts` here: https://github.com/PennyLaneAI/pennylane/blob/b904c966d124e66dbf82cc0b42f580db2d9a7cc1/pennylane/_device.py#L732
- Should we add `AllCounts` here? https://github.com/PennyLaneAI/pennylane/blob/b904c966d124e66dbf82cc0b42f580db2d9a7cc1/pennylane/devices/default_mixed.py#L580
- Add `AllCounts` in these two lines:
https://github.com/PennyLaneAI/pennylane/blob/b904c966d124e66dbf82cc0b42f580db2d9a7cc1/pennylane/tape/tape.py#L481
https://github.com/PennyLaneAI/pennylane/blob/b904c966d124e66dbf82cc0b42f580db2d9a7cc1/pennylane/tape/tape.py#L482

File: doc/introduction/measurements.rst
Comment (AlbertMitjans) [inline doc/introduction/measurements.rst:184]: You need to indent the code block.

File: pennylane/_qubit_device.py
Comment (AlbertMitjans) [inline pennylane/_qubit_device.py:1561]: Add test to cover this.

File: pennylane/_qubit_device.py
Comment (AlbertMitjans) [reply pennylane/_qubit_device.py:1564]: We can leave it don't worry.

File: pennylane/_qubit_device.py
Comment (lillian542) [reply pennylane/_qubit_device.py:1561]: This was added to deal with an `EigenvalsUndefinedError` for `qml.Hermitian`, but it seems the error is no longer raised now that the eigenvalues are found using `outcomes = qml.eigvals(obs)` instead of `outcomes = obs.compute_eigvals()`. At least I haven't been able to find any instances now where this exception is relevant, so I've taken it out.

**Comment by lillian542** [comment]
@AlbertMitjans Sorry for the delay, its been a busy few days. I think everything is ready to go now.

File: tests/test_measurements.py
Comment (AlbertMitjans) [inline tests/test_measurements.py:949]: Maybe it is better to
```suggestion
        assert res[1] == {"00": 10}
        assert res[2] == {"00": 10, "01": 0, "10": 0, "11": 0}
```

**Review by AlbertMitjans — COMMENTED** [review]
Great job! 💯 

I'll approve it once these small changes are applied!

File: pennylane/_qubit_device.py
Comment (AlbertMitjans) [inline pennylane/_qubit_device.py:1623]: Indent the code block

File: pennylane/tape/tape.py
Comment (AlbertMitjans) [inline pennylane/tape/tape.py:482]: Lines too long

**Comment by lillian542** [comment]
Changes done! @AlbertMitjans

File: pennylane/_qubit_device.py
Comment (AlbertMitjans) [inline pennylane/_qubit_device.py:1638]:
```suggestion
            outcomes = qml.eigvals(obs)
```

File: pennylane/measurements.py
Comment (AlbertMitjans) [inline pennylane/measurements.py:803]: Indent block

File: pennylane/measurements.py
Comment (AlbertMitjans) [inline pennylane/measurements.py:815]: Use `.. code-block:: python3` and indent block

**Review by AlbertMitjans — COMMENTED** [review]
Some more doc changes

**Comment by antalszava** [comment]
Hi @lillian542, double-checking how is it going here? Could we help in some way to push for merging?

**Comment by lillian542** [comment]
@antalszava @AlbertMitjans I'm sorry for my slow response, you've caught me at an extremely busy time. I've made the most recently suggested changes. 

I'm not sure how to set the checks to run to make sure there are no further issues that need to be addressed. Can I do that on my end, and if so, how? 

I'll make sure to prioritise responding quickly so we can get this wrapped up.

**Comment by antalszava** [comment]
> @antalszava @AlbertMitjans I'm sorry for my slow response, you've caught me at an extremely busy time. I've made the most recently suggested changes.

> I'm not sure how to set the checks to run to make sure there are no further issues that need to be addressed. Can I do that on my end, and if so, how?

No worries there! Let us know if we can help out even with development. :)

Just merged `master` in and it seems that only the formatting check is off.

Otherwise, the tests can be run locally via `pytest tests`, see more here: https://docs.pennylane.ai/en/stable/development/guide/tests.html.
And see our documentation section on running the CI checks: https://docs.pennylane.ai/en/stable/development/guide/pullrequests.html#continuous-integration-checks

Let us know if we can help further!

**Comment by AlbertMitjans** [comment]
Hi @lillian542 , as mentioned in the github action, the file `pennylane/tape/tape.py` needs reformatting. To do so please run (make sure to install `black` first: `pip install black`):
```
black -l 100 pennylane/
```
Finally commit the changes.

Hope this helps. :)

**Review by AlbertMitjans — APPROVED** [review]
Good job!! 🚀

**Pull request merged**
